### ISSUE TITLE
AI: Checking possible target combinations takes far too long

### ISSUE THREAD

**Issue by Alex-Vasile** [magefree/mage#9539]
The issue is straightforward to describe: Many cards have targeted triggered abilities, e.g. "When {A} {do B to} target ..." or worse "when {A} {do B to} **up to**...". These sorts of cards really slow down an AI game.

See previously reported issues with [[Deepglow Skate]] in #9438 and [[Elder Deep-Fiend]] in #9518.

They are especially bad if the AI has them since it will keep checking all possible targets whenever it gets priority. The number of options for it to check can explode.

However, they will slow down the game even if the human player is using these cards.

Further something strange is going on in terms of when this slowdown is occuring. For example, If you have 30+ hands and Deepglow Skate in your hand, you will be able to cast the spell without issue. The slowdown will only occur AFTER the spell has resolved and the player is choosing targets.
(This strange behaviour may be AI difficulty setting related, perhaps on lower levels it does not think far enough ahead for it to get to the point of evaluating all possible target combinations).

Either way, I proposed two change to deal with these issues:
1. For spells like Elder Deep-Fiend, or for abilities like Deepglow Skate (once it's entered and its ability triggered) the AI should do **nothing** until the player has chosen targets. There is no point for the AI to guess what to do as checking all of the combinations takes too long, and it cannot respond until after the player has made their choices anyway.
2. When it comes to the AI's turn to cast such spells, or when looking ahead to see if it should counter the Deepglow Skate, it should use a worst-case scenario heuristic (the player who has to make the choice will make the choice that puts them furthest ahead and hampers each of their enemies).

Or if 2 cannot be implemented generally, then some more serious pruning must be applied to the decisions that the AI evaluates.

**Comment by Alex-Vasile** [comment]
If the server has enough memory this will slow down the game, but it can cause out of memory errors. see #6979

**Comment by PurpleCrowbar** [comment]
@JayDi85 Does #13892 fix this?

**Comment by JayDi85** [comment]
@PurpleCrowbar nope, I don't think so. It’s old issue from 2022 and master has many related AI fixes after that: memory leaks on AI sims, same targets optimization, less max actions chain, better timeout processing, etc.

P.S. I’ll look at use cases from that issue later — is it actual anymore.

**Comment by JayDi85** [comment]
Well, AI code already had some possible targets optimizations, but it was used for `TargetAmount` only. I added same optimizations for `TargetImpl` too (see 2833460e593bd4cf61aba8bb8a5ef4d1c07ed75b). So both use cases with 100% CPU and server crashes are fine now (Deepglow Skate, Elder Deep-Fiend, etc).

It's impossible to fix the [combinatorial explosion problem](https://en.wikipedia.org/wiki/Combinatorial_explosion) (more targets, more combinations, more resources). But it allow to increase max limits and add more stability to AI. Current targets optimization algorithm combine targets by same groups and calculate min possible targets list from it (see `TargetOptimization.java`).

As example: if you must target x1 land from x10 forests and x10 mountains -- optimization will keep only unique options -- x1 forest and x1 mountain, e.g. x2 possible targets instead x20. Same for multiple targets. If you have too much unique targets on battlefield then you will catch performance issues anyway -- AI will do nothing on timeout.

Unique card characteristics to group:
<img width="720" height="239" alt="Image" src="https://github.com/user-attachments/assets/3134e426-2699-470c-90d1-cdba79b3fd2d" />

Unique permanent characteristics to group:
<img width="760" height="380" alt="Image" src="https://github.com/user-attachments/assets/819aa1d2-1b22-4892-9832-03152f1a70f2" />

**Comment by PurpleCrowbar** [comment]
Just had the server crash after playing Deepglow Skate. This message was in the console:

```
WARN  2025-10-21 18:14:43,724 AI player thinks too long (report it to github):                                           =>[GAME 947fd4e3-031e-4590-94ce-4d0335aca80b] ComputerPlayer6.addActionsTimed 
WARN  2025-10-21 18:14:43,724  - player: Computer 2                                                                      =>[GAME 947fd4e3-031e-4590-94ce-4d0335aca80b] ComputerPlayer6.addActionsTimed 
WARN  2025-10-21 18:14:43,724  - battlefield size: 40                                                                    =>[GAME 947fd4e3-031e-4590-94ce-4d0335aca80b] ComputerPlayer6.addActionsTimed 
WARN  2025-10-21 18:14:43,724  - stack: 1 (top: stack ability (When {this} enters, each opponent sacrifices a nontoken creature or planeswalker o...) =>[GAME 947fd4e3-031e-4590-94ce-4d0335aca80b] ComputerPlayer6.addActionsTimed 
WARN  2025-10-21 18:14:43,724  - game: !!!SIMULATION, AI!!! Commander Free For All; T34.M1; active: PurpleCrowbar; stack: 1 (top: stack ability (When {this} enters, each opponent sacrifices a nontoken creature or planeswalker o...) =>[GAME 947fd4e3-031e-4590-94ce-4d0335aca80b] ComputerPlayer6.addActionsTimed 
WARN  2025-10-21 18:14:43,724 Possible freeze chain:                                                                     =>[GAME 947fd4e3-031e-4590-94ce-4d0335aca80b] ComputerPlayer6.printFreezeNode 
WARN  2025-10-21 18:14:43,724  - unknown use case (too many possible targets?)                                           =>[GAME 947fd4e3-031e-4590-94ce-4d0335aca80b] ComputerPlayer6.printFreezeNode 
```

Seems the AI still explodes when an ability has too many possible targets

**Comment by JayDi85** [comment]
Yes, it's will freeze due too many diff targets: "battlefield size: 40". 

Maybe better optimization can be used here like:
- try to use all unique characteristics to group same targets (see screenshot from https://github.com/magefree/mage/issues/9539#issuecomment-3172147519 );
- if it still too big then remove some characteristics, e.g. keep only owner, name and PT;
- if it still too big then generate random groups (AI will miss some targets but it's better compared to freeze/crash);

The main idea -- target optimization must return only few possible targets for faster calc.

**Comment by PurpleCrowbar** [comment]
Re-opening this as the target optimisation commit I made reduces the frequency of the issue but doesn't solve it. It's still possible (and likely) that a user would encounter this bug in the future.
